File: web/src/services/imagegate.ts
import sharp from 'sharp';
import type { Config } from '../config.js';

export interface GateResult {
  ok: true;
  webp: Buffer;
  width: number;
  height: number;
}
export interface GateReject {
  ok: false;
  reason: string;
}

const JPEG_MAGIC = Buffer.from([0xff, 0xd8, 0xff]);
const PNG_MAGIC = Buffer.from([0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a]);

function hasMagic(bytes: Buffer): boolean {
  if (bytes.length >= 3 && bytes.subarray(0, 3).equals(JPEG_MAGIC)) return true;
  if (bytes.length >= 8 && bytes.subarray(0, 8).equals(PNG_MAGIC)) return true;
  return false;
}

export async function gateImage(
  bytes: Buffer,
  cfg: Config,
): Promise<GateResult | GateReject> {
  // 1. Byte-length cap FIRST — cheapest, before any decode work.
  if (bytes.length > cfg.quotas.max_photo_bytes) {
    return { ok: false, reason: 'file too large' };
  }

  // 2. Magic bytes — JPEG/PNG only in M2 (HEIC is M4). Reject anything else
  //    before sharp ever touches the buffer.
  if (!hasMagic(bytes)) {
    return { ok: false, reason: 'unsupported format' };
  }

  // 3. Decode + re-encode. limitInputPixels enforces the pixel cap (sharp
  //    throws when exceeded). .rotate() applies EXIF orientation; the WebP
  //    re-encode strips all metadata (EXIF/orientation gone). failOn:'error'
  //    makes truncated/corrupt inputs throw rather than silently pass.
  try {
    const pipeline = sharp(bytes, {
      limitInputPixels: cfg.quotas.max_pixels,
      failOn: 'error',
    })
      .rotate()
      .webp({ quality: 75 });

    const { data, info } = await pipeline.toBuffer({ resolveWithObject: true });
    return { ok: true, webp: data, width: info.width, height: info.height };
  } catch (err) {
    // sharp's pixel-limit error message contains "pixels"; distinguish it
    // from a generic decode failure so callers/tests get a precise reason.
    const msg = err instanceof Error ? err.message.toLowerCase() : '';
    if (msg.includes('pixel')) {
      return { ok: false, reason: 'image too large' };
    }
    return { ok: false, reason: 'corrupt image' };
  }
}

File: worker/notbulk/imagegate.py
"""cv2-based image gate for FETCHED bytes (worker side), mirroring the Node
sharp gate for uploads. Enforces the SAME limits from the SAME config keys:
magic bytes (JPEG/PNG), byte-size cap, decode, pixel cap, re-encode to WebP q75.
"""
from __future__ import annotations

import cv2
import numpy as np

_JPEG_MAGIC = b"\xff\xd8\xff"
_PNG_MAGIC = b"\x89PNG\r\n\x1a\n"


class GateRejected(Exception):
    """Permanent rejection of fetched bytes."""


def gate_bytes(data: bytes, cfg: dict) -> tuple[bytes, int, int]:
    """Return (webp_bytes, width, height) or raise GateRejected."""
    max_bytes = int(cfg["quotas"]["max_photo_bytes"])
    max_pixels = int(cfg["quotas"]["max_pixels"])

    if len(data) > max_bytes:
        raise GateRejected("image too large (bytes)")
    if not (data.startswith(_JPEG_MAGIC) or data.startswith(_PNG_MAGIC)):
        raise GateRejected("bad magic bytes (JPEG/PNG only)")

    img = cv2.imdecode(np.frombuffer(data, np.uint8), cv2.IMREAD_COLOR)
    if img is None:
        raise GateRejected("decode failed")
    h, w = img.shape[:2]
    if h * w > max_pixels:
        raise GateRejected("pixel cap exceeded")

    ok, buf = cv2.imencode(".webp", img, [cv2.IMWRITE_WEBP_QUALITY, 75])
    if not ok:
        raise GateRejected("webp re-encode failed")
    return buf.tobytes(), w, h
